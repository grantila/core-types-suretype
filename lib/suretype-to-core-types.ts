import type { ConversionResult, NodeDocument } from 'core-types'
import { convertJsonSchemaToCoreTypes } from 'core-types-json-schema'

import { convertSuretypeToJsonSchema } from './suretype-to-json-schema.js'
import type { SuretypeToJsonSchemaOptions } from './types.js'


export async function convertSureTypeToCoreTypes(
	filename: string,
	options: SuretypeToJsonSchemaOptions = { }
)
: Promise< ConversionResult< NodeDocument > >
{
	const resSchema = await convertSuretypeToJsonSchema( filename, options );
	const { data: jsonSchema } = resSchema;

	const resCoreTypes = convertJsonSchemaToCoreTypes( jsonSchema );

	return {
		convertedTypes: resCoreTypes.convertedTypes,
		notConvertedTypes: [ ...new Set( [
			...resSchema.notConvertedTypes,
			...resCoreTypes.notConvertedTypes,
		] ) ],
		data: resCoreTypes.data,
	};
}
