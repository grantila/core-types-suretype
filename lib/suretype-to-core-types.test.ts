import path from 'node:path'
import { fileURLToPath } from 'node:url'

import { convertSureTypeToCoreTypes } from './suretype-to-core-types.js'


const __dirname = path.dirname( fileURLToPath( import.meta.url ) );

describe( "convertSuretypeToJsonSchema", ( ) =>
{
	it( "should convert types with dependencies", async ( ) =>
	{
		const file = path.join( __dirname, 'test', 'validator.js' );
		const res = await convertSureTypeToCoreTypes( file );
		const { convertedTypes, notConvertedTypes, data } = res;

		expect( convertedTypes.sort( ) )
			.toStrictEqual( [ 'ChatLine', 'User' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );

		( data as any ).types.sort( ( a: any, b: any ) =>
			a.name.localeCompare( b.name )
		);

		expect( data ).toMatchSnapshot( );
	} );
} );
