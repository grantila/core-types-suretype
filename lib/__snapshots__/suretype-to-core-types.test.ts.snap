// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertSuretypeToJsonSchema should convert types with dependencies 1`] = `
{
  "types": [
    {
      "additionalProperties": true,
      "name": "ChatLine",
      "properties": {
        "line": {
          "node": {
            "type": "string",
          },
          "required": true,
        },
        "user": {
          "node": {
            "ref": "User",
            "type": "ref",
          },
          "required": false,
        },
      },
      "title": "A chat line",
      "type": "object",
    },
    {
      "additionalProperties": true,
      "description": "A user representation",
      "examples": "[
  "{ firstName: \\"joe\\", id: \\"12345678\\" }"
]",
      "name": "User",
      "properties": {
        "firstName": {
          "node": {
            "type": "string",
          },
          "required": true,
        },
        "id": {
          "node": {
            "type": "string",
          },
          "required": true,
        },
      },
      "title": "The user type",
      "type": "object",
    },
  ],
  "version": 1,
}
`;
